File: software/Filter.py
#!/usr/bin/python
# -*- coding: utf-8 -*
#
# This file is part of SillyCat Development Tools.
#
# SillyCat Development Tools is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.

# SillyCat Development Tools is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public License
# along with SillyCat Development Tools.  If not, see <http://www.gnu.org/licenses/>.

__author__ = 'andreas.dahlberg90@gmail.com (Andreas Dahlberg)'
__version__ = '0.1.0'


def begins_with(stream_data, keyword):
	return stream_data.startswith(keyword)

def ends_with(stream_data, keyword):
	return stream_data.endswith(keyword)	

def contains(stream_data, keyword):
	return stream_data.count(keyword) > 0



class Filter():
	def __init__(self):
		self._active_ = []


	def filter(stream_data):
		pass


	def add_start_keyword(keyword):
		pass






if __name__ == '__main__':
	print(contains('spam end eggs', 'end'))

File: firmware/src/utility/Filter/Filter.c
/**
 * @file   Filter.c
 * @Author Andreas Dahlberg (andreas.dahlberg90@gmail.com)
 * @date   2018-02-11 (Last edit)
 * @brief  Implementation of an exponential moving average(EMA) filter.
 *
 * Detailed description of file.
 */

/*
This file is part of SillyCat firmware.

SillyCat firmware is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

SillyCat firmware is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with SillyCat firmware.  If not, see <http://www.gnu.org/licenses/>.
*/

//////////////////////////////////////////////////////////////////////////
//INCLUDES
//////////////////////////////////////////////////////////////////////////

#include "Filter.h"
#include "sc_assert.h"

//////////////////////////////////////////////////////////////////////////
//DEFINES
//////////////////////////////////////////////////////////////////////////

//////////////////////////////////////////////////////////////////////////
//TYPE DEFINITIONS
//////////////////////////////////////////////////////////////////////////

//////////////////////////////////////////////////////////////////////////
//VARIABLES
//////////////////////////////////////////////////////////////////////////

//////////////////////////////////////////////////////////////////////////
//LOCAL FUNCTION PROTOTYPES
//////////////////////////////////////////////////////////////////////////

//////////////////////////////////////////////////////////////////////////
//FUNCTIONS
//////////////////////////////////////////////////////////////////////////

void Filter_Init(struct filter_t *self_p, int16_t initial_value, uint16_t alpha)
{
    sc_assert(self_p != NULL);

    self_p->value = initial_value;
    self_p->alpha = alpha;
}

void Filter_Process(struct filter_t *self_p, int16_t sample)
{
    sc_assert(self_p != NULL);

    /* Use 32-bits to prevent overflow. */
    uint32_t tmp = (int32_t)self_p->alpha * (int32_t)sample +
                   (int32_t)(65536 - self_p->alpha) * (int32_t)self_p->value;

    /* Scale back to 16-bit. */
    self_p->value = (int16_t)((tmp + 32768) / 65536);
}

int16_t Filter_Output(const struct filter_t *self_p)
{
    sc_assert(self_p != NULL);

    return self_p->value;
}

bool Filter_IsInitialized(const struct filter_t *self_p)
{
    sc_assert(self_p != NULL);

    return self_p->alpha != 0;
}

//////////////////////////////////////////////////////////////////////////
//LOCAL FUNCTIONS
//////////////////////////////////////////////////////////////////////////
